### DISCUSSION TITLE
[Idea] Prevent BO notifications from slowing the shop down

### DISCUSSION THREAD

**Discussion by cricket-hockey** [PrestaShop/PrestaShop#32279] [Ideas]
(edited by @eternoendless to summarize)

#### The problem

Since upgrading to 1.7.7.2 the orders page in the backend, and clicking into any order is incredibly slow. Most of the time it takes around 60 seconds for the orders page to load and then another 60 seconds when you click into an order. Occasionally it will load in around 5 seconds but 90% of the time it takes around 1 minute.

I never had this problem on my previous version which was 1.7.6.2

#### The solution

(proposed in [this comment](https://github.com/PrestaShop/PrestaShop/issues/23827#issuecomment-871417476))

1. Enable Back Office user to turn off BO notifications
2. When PrestaShop fetches the data for BO notifications, track the time of the data fetching. If it's above one threshold (say, 10 seconds) we disable BO notifications and create a warning message in logs.
3. Make sure that BO notifications is loaded asynchronously so that even though it consumes a lot of resources ; it does not slow the opening of a BO page

#### Steps to Reproduce

1. Go to 'Orders' tab in back office and click on Orders

**Screenshots**

See attached screenshot of when profiling is enabled:

![screenshot-wwwcricket-hockeycom-20210324-12_05_33](https://user-images.githubusercontent.com/81647673/113041904-6cb53600-9192-11eb-8420-d89d1359d50a.png)

As you can see config is the issue. I have uninstalled all third party modules and it makes no difference. Our hosting company has been through everything and done a lot of testing and are convinced it is nothing to do with the server. They have found this query which takes about 25 seconds to return 5 rows. It's reading over 100,000 rows to get 5 out !

SELECT SQL_CALC_FOUND_ROWS o.id_order, o.id_customer, o.total_paid, o.id_currency, o.date_upd, c.firstname, c.lastname, ca.name, co.iso_code
FROM ps_orders as o
LEFT JOIN ps_customer as c ON (c.id_customer = o.id_customer)
LEFT JOIN ps_carrier as ca ON (ca.id_carrier = o.id_carrier)
LEFT JOIN ps_address as a ON (a.id_address = o.id_address_delivery)
LEFT JOIN ps_country as co ON (co.id_country = a.id_country)
WHERE id_order > 0 AND o.id_shop IN (1)
ORDER BY id_order DESC
LIMIT 5

Could this be the problem?

#### Additional information

* PrestaShop version: 1.7.7.2
* PHP version: 7.3.26
* 200,000+ orders and customers - is this anything to do with it? Should I delete the olders ones to bring this down? However, this was never a problem on the previous version.

**Comment by kpodemski** [comment]
Hey @cricket-hockey 

can you let us know how many records do you have in `PREFIX_configuration` table?

You gave us screenshot from profiler but unfortunately there's nothing about SQL queries and this 25s... I remember this kind of issue and it was related to static cache and too much of configuration records.

**Comment by cricket-hockey** [comment]
Hi @kpodemski 

Thanks for your reply. There are 24823 records in my configuration table, which presumably is too high. Is there anything I can do about this?

**Comment by kpodemski** [comment]
It's not that many, but of course you should check for duplicates in `ps_configuration` table, I remember some scenarios where one of the popular newsletter modules used this table to store some data related to customer segmentation.

**Comment by cricket-hockey** [comment]
@kpodemski OK, thanks, will do.

What about that query I posted? Is that not part of the problem?

Another thing I have found is that when just one person is logged into Prestashop then the speed is not too bad. But as soon as other users login, the orders page and going in to any order takes up to 1 minute to load.

**Comment by kpodemski** [comment]
@cricket-hockey I think it might have something to do with your hosting, not PrestaShop itself, there is something wrong with `configuration` part as you can see in the profiler.

About query you posted - it's based on indexes so it shouldn't be a problem and profiler shows that querying took only small amount of time.

**Comment by cricket-hockey** [comment]
@kpodemski The only thing that makes me think it isn't the hosting is that this only started when I upgraded to 1.7.7.2. If it was the hosting, would this not have happened when I was on 1.7.6.x? They have been very good and spent weeks looking into everything - even copying the site to another server and running tests there. Nothing had any impact.

**Comment by cricket-hockey** [comment]
Almost all of the configuration table is filled with entries like this:

![screenshot-89 145 84 90-2021 03 31-08_26_25](https://user-images.githubusercontent.com/81647673/113106647-eaad2780-91fa-11eb-9912-c8b18e78d5e1.png)

This is from my SEO URL module. Could this be the problem?

It's odd though that the problem only affects the backend orders page. If it was this or the hosting, surely the rest of Prestashop and the front end would be slow too?

**Comment by kpodemski** [comment]
@cricket-hockey 

thing is that SQL queries aren't slow from what you show us in screenshot from profiler, this is why it's weird - also, it's surprising that occasionally this page load quicker 🤔 

For now it's hard for me to say if it's related to new Orders page, I don't have any bigger shops updated to 1.7.7 just yet but I'll try to generate some orders and see how it's working for me.

**Comment by hibatallahAouadni** [comment]
Hello @cricket-hockey 

Could you please check this issue #14703 it seems similar to yours.
Please check and feedback.

PS: thanks @kpodemski for your quick replies :rocket: 

Thanks!

**Comment by cricket-hockey** [comment]
@hibatallahAouadni Thanks for your reply. I have checked that issue but unless I am missing something, there doesn't seem to be a fix on there for it? Someone just commented out the two queries. Should I try that?

**Comment by hibatallahAouadni** [comment]
@cricket-hockey yep try it and see if it could help you up. Don't forget to feedback :wink:

**Comment by cricket-hockey** [comment]
@hibatallahAouadni OK, I have commented out those 2 queries but it is still the same.

I've switched profiling on again and here is the result when clicking into the orders page. I thought the fact that config is the problem indicates this is not due to a query?

[Untitled 1.pdf](https://github.com/PrestaShop/PrestaShop/files/6242720/Untitled.1.pdf)

**Comment by cricket-hockey** [comment]
@hibatallahAouadni @kpodemski Do you think this could be anything to do with languages? Generally most sections on the backend are slower than they should be, as well as the website. In the profiling I'm seeing a lot of slow loading times with anything related to languages. I have 9 different languages so maybe this is part of the problem?

**Comment by gennaris** [comment]
@kpodemski @cricket-hockey 
I am facing the same problem during a 1.6 => 1.7 data migration of a huge website (about 900k+ orders in **ps_orders**).
In my case target site is based on 1.7.6.2 version.

The same reported query from OP :

```
SELECT SQL_CALC_FOUND_ROWS o.`id_order`, o.`id_customer`, o.`total_paid`, o.`id_currency`, o.`date_upd`, c.`firstname`, c.`lastname`, ca.`name`, co.`iso_code`
                                        FROM `ps_orders` as o
                                        LEFT JOIN `ps_customer` as c ON (c.`id_customer` = o.`id_customer`)
                                        LEFT JOIN `ps_carrier` as ca ON (ca.`id_carrier` = o.`id_carrier`)
                                        LEFT JOIN `ps_address` as a ON (a.`id_address` = o.`id_address_delivery`)
                                        LEFT JOIN `ps_country` as co ON (co.`id_country` = a.`id_country`)
                                        WHERE `id_order` > 124 AND o.id_shop IN (1)
                                        ORDER BY `id_order` DESC
                                        LIMIT 5 
```
is taking 40+ seconds on my staging box (4 core VPS) and it seems is launched both on first opening
of the Products and Order page in backoffice making it almost unusable.

My ps_configuration table have 5k rows by the way, 
anyway it's not involved in that query so it shouldn't be the culprit here.

Anybody knows the purpose of this query and where it is called in the core so I can investigate further and maybe provide more useful information... ?

**Comment by hibatallahAouadni** [comment]
Hello @cricket-hockey 

> Thanks for your reply. I have checked that issue but unless I am missing something, there doesn't seem to be a fix on there for it?

That issue is not supposed to have some fix, I asked you if it's similar to yours. So, could you please check it again #14703.

> I am facing the same problem during a 1.6 => 1.7 data migration of a huge website (about 900k+ orders in ps_orders).
In my case target site is based on 1.7.6.2 version.

@kpodemski any idea about this issue? :sweat_smile: 

Thanks!

**Comment by gennaris** [comment]
Hello all,
FYI - I've done some investigation on that query. 

It's called in **/classes/Notification.php** , inside _getLastElementsIdsByType_ method for notifications refresh.
Since query is using SQL_CALC_FOUND_ROWS and join all "critical" ps tables - It runs for several seconds , over-stressing DB on huge populated shops.

I had to override the method to the customer , otherwise the backoffice was unusable...

**Comment by cricket-hockey** [comment]
@gennaris This sounds like exactly the same problem I am having.

Would you be able to give me the override you used please?

**Comment by gennaris** [comment]
@cricket-hockey sure - note that this will cause loss of notifications in bo.

Inside the **getLastElements()** method :

Add this :

        foreach($this->types as $type) {
            $notifications[$type] = array('total' => 0, 'results' => array());
        }
 
        return $notifications;

Right after 
    
    $notifications = [];
    
This way the queries inside **getLastElementsIdsByType()** to calculate notifications will never be triggered.

**Comment by cricket-hockey** [comment]
@gennaris Thank you, you're a star!

We don't need the BO notifications, so that's no problem. My code (1.7.7.2) was a little different though:

`    public function getLastElements()
    {
        $notifications = [];
        $employeeInfos = Db::getInstance(_PS_USE_SQL_SLAVE_)->getRow('
		SELECT id_last_order, id_last_customer_message, id_last_customer
		FROM `' . _DB_PREFIX_ . 'employee`
		WHERE `id_employee` = ' . (int) Context::getContext()->employee->id);

        foreach ($this->types as $type) {
            $notifications[$type] = Notification::getLastElementsIdsByType($type, $employeeInfos['id_last_' . $type]);
        }

        return $notifications;
    }`

....so I just commented out the whole lot and the problem was instantly solved. Now any page in the BO only takes a few seconds to load rather than up to 1 minute.

Is it OK to leave this commented out? It won't have any other knock on effects, will it?

@hibatallahAouadni @kpodemski Just tagging you for reference so you can see where the problem was.

**Comment by gennaris** [comment]
@cricket-hockey 

Adding the code after : 

    $notifications = []; 

will be fine.

Not sure if commenting everything will trigger PHP warning / errors here and there.

I edited the function to return the same PS expected array with 0 values to avoid triggering the "problematic" queries.

**Comment by hibatallahAouadni** [comment]
Hello everyone,

Thanks @gennaris for your quick replies, your super helpful solutions and your detailed explanation :wink: 
So, to summarize the issue:

> It's about BO notifications icons counters that needed to be refreshed **automatically**
Well, the issue occurs only with huge databases (lot of orders), the query responsible to refresh the counters takes a lot of time / resources. So, it blocks the whole shop :confounded: 

So, I will relabel this issue from **Bug** to **Improvement** cause it's a performance issue :relaxed: 
The Product Team will take it into consideration for future developments.

Please be aware that there is no guarantee that this feature will be developed anytime soon. If this is important to you, we strongly suggest you consider [hiring a specialist](https://www.prestashop.com/en/experts) to help you create it. 

If you are willing to implement this feature into PrestaShop as a [Pull Request](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/), the teams will be happy to help you by ensuring that it is appropriate and coherent with the rest of the project. If that is the case, please let us know! 

Thanks!

**Comment by matks** [comment]
Hello everyone I tried to improve the bug report title to match the issue. Please tell me if it does not look right.

So, I can think of a few easy ways to ease the problem
1. Enable Back Office user to turn off BO notifications
2. When PrestaShop fetches the data for BO notifications, track the time of the data fetching. If it's above one threshold (say, 10 seconds) we disable BO notifications and create a warning message in logs.
3. Make sure that BO notifications is loaded asynchronously so that even though it consumes a lot of resources ; it does not slow the opening of a BO page

**Comment by cricket-hockey** [comment]
@matks Sounds good to me

**Reply by devgraal** [reply]
We had the exact same problem on our side.
The solution had been to modify the SQL Query, and so the Notification class.
The reason the query is that long on huge database is the "SQL_CALC_FOUND_ROWS"
Doing the query without this, and one more with only a "count(*)" fixed the very slow execution
Of course the "count" part can be optimized ^^
For stats:
- before modification, query exec in approx 50sec
- after modification, with the 2 queries, less than 1 sec

**Reply by eternoendless** [reply]
@devgraal Please consider sending a Pull Request with the modification!
